gstaichi/python/dlpack_funcs.cpp

```diff
@@ -11,90 +11,204 @@
 
 namespace gstaichi {
 namespace lang {
-pybind11::capsule ndarray_to_dlpack(Program *program,
-                                    pybind11::object owner,
-                                    Ndarray *ndarray) {
-  auto *owner_holder = new pybind11::object(owner);
 
-  DeviceAllocation devalloc = ndarray->get_device_allocation();
+void validate_arch(Arch arch) {
```

Comment:
this git diff, in github ui, is very unreadable. Here's what I actually did:
- created a new function, field_to_dlpack, that is above the old function, ndarray_to_dlpack
- factorized out get_raw_ptr
- factorized out get_type_info